WTG1001:DoNotUsePrivateKeyword should trigger on 'private partial void' method and not trigger 'private partial void' method with an 'out' keyword

To resolve the conflict between `CS8796: Partial method must have accessibility modifiers because it has a non-void return type` and `WTG1001:Do not use the 'private' keyword`, latest WTG.Analyzer have 'private partial' method to be ignored by WTG1001. However, the two cases are currently not covered, examples of which are:
```
private partial void Foo(); // -> currently no error but should trigger WTG1001 and be 'partial void Foo();'
private partial void Foo(out int outValue); // -> satisfy WTG1001 and CS8796, no changes required
```
Initial issue was fixed by #205 
Extended from the issue #220 

Corresponds to [WI00764889](https://svc-ediprod.wtg.zone/Services/edit/WI/WI00764889)

For reference the error avoided here by specifying `private` on the second line is:

> error CS8797: Partial method 'Foo(out int)' must have accessibility modifiers because it has 'out' parameters.

--- a/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Diagnostics.xml
+++ b/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Diagnostics.xml
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8" ?>
 <diagnostics severity="Hidden">
 	<languageVersion>9.0</languageVersion>
-	<suppressId justtification="To ignore when only one method in one class have the code fix removing 'private' applied.">CS8799</suppressId>
+	<suppressId>CS8799</suppressId>
 	<diagnostic id="WTG1001" message="Our convention is to omit the 'private' modifier where it is already the default.">
 		<location>Test0.cs: (9,3-10)</location>
 	</diagnostic>

--- a/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Diagnostics.xml
+++ b/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Diagnostics.xml
@@ -1,10 +1,17 @@
 <?xml version="1.0" encoding="utf-8" ?>
 <diagnostics severity="Hidden">
 	<languageVersion>9.0</languageVersion>
+	<suppressId justtification="To ignore when only one method in one class have the code fix removing 'private' applied.">CS8799</suppressId>
 	<diagnostic id="WTG1001" message="Our convention is to omit the 'private' modifier where it is already the default.">
-		<location>Test0.cs: (15,3-10)</location>
+		<location>Test0.cs: (9,3-10)</location>
+	</diagnostic>
+	<diagnostic id="WTG1001" message="Our convention is to omit the 'private' modifier where it is already the default.">
+		<location>Test0.cs: (19,3-10)</location>
+	</diagnostic>
+	<diagnostic id="WTG1001" message="Our convention is to omit the 'private' modifier where it is already the default.">
+		<location>Test0.cs: (25,3-10)</location>
 	</diagnostic>
 	<diagnostic id="WTG1006" message="Our convention is to omit the 'internal' modifier on types where it is already the default." >
-		<location>Test0.cs: (20,2-10)</location>
+		<location>Test0.cs: (30,2-10)</location>
 	</diagnostic>
 </diagnostics>

--- a/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Result.cs
+++ b/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Result.cs
@@ -3,11 +3,19 @@ namespace ns
 	partial class Foo
 	{
 		private partial int Bar();
+		public partial void FooBar();
+
+		partial void Qux();
+		private partial void Quux(out int value);
 	}
 
 	partial class Foo
 	{
 		private partial int Bar() { return default; }
+		public partial void FooBar() { }
+
+		partial void Qux() { }
+		private partial void Quux(out int value) { value = default; }
 	}
 
 	class Outer

--- a/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Result.cs
+++ b/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Result.cs
@@ -14,7 +14,7 @@ namespace ns
 	{
 		private partial int Bar() { return default; }
 		public partial void FooBar() { }
-		private partial int FooBarBaz(out int value) { value = default; return default; }
+		private partial int FooBarBaz(out int value) => throw null;
 
 		partial void Qux() { }
 		private partial void Quux(out int value) { value = default; }

--- a/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Result.cs
+++ b/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Result.cs
@@ -4,6 +4,7 @@ namespace ns
 	{
 		private partial int Bar();
 		public partial void FooBar();
+		private partial int FooBarBaz(out int value);
 
 		partial void Qux();
 		private partial void Quux(out int value);
@@ -13,6 +14,7 @@ namespace ns
 	{
 		private partial int Bar() { return default; }
 		public partial void FooBar() { }
+		private partial int FooBarBaz(out int value) { value = default; return default; }
 
 		partial void Qux() { }
 		private partial void Quux(out int value) { value = default; }

--- a/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Source.cs
+++ b/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Source.cs
@@ -4,6 +4,7 @@ namespace ns
 	{
 		private partial int Bar();
 		public partial void FooBar();
+		private partial int FooBarBaz(out int value);
 
 		private partial void Qux();
 		private partial void Quux(out int value);
@@ -13,6 +14,7 @@ namespace ns
 	{
 		private partial int Bar() { return default; }
 		public partial void FooBar() { }
+		private partial int FooBarBaz(out int value) { value = default; return default; }
 
 		private partial void Qux() { }
 		private partial void Quux(out int value) { value = default; }

--- a/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Source.cs
+++ b/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Source.cs
@@ -14,7 +14,7 @@ namespace ns
 	{
 		private partial int Bar() { return default; }
 		public partial void FooBar() { }
-		private partial int FooBarBaz(out int value) { value = default; return default; }
+		private partial int FooBarBaz(out int value) => throw null;
 
 		private partial void Qux() { }
 		private partial void Quux(out int value) { value = default; }

--- a/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Source.cs
+++ b/src/WTG.Analyzers.Test/TestData/VisibilityAnalyzer/Partial/Source.cs
@@ -3,11 +3,19 @@ namespace ns
 	partial class Foo
 	{
 		private partial int Bar();
+		public partial void FooBar();
+
+		private partial void Qux();
+		private partial void Quux(out int value);
 	}
 
 	partial class Foo
 	{
 		private partial int Bar() { return default; }
+		public partial void FooBar() { }
+
+		private partial void Qux() { }
+		private partial void Quux(out int value) { value = default; }
 	}
 
 	class Outer

--- a/src/WTG.Analyzers/Analyzers/Visibility/VisibilityAnalyzer.cs
+++ b/src/WTG.Analyzers/Analyzers/Visibility/VisibilityAnalyzer.cs
@@ -50,7 +50,7 @@ namespace WTG.Analyzers
 
 					case SyntaxKind.ProtectedKeyword:
 					case SyntaxKind.PublicKeyword:
-					case SyntaxKind.PartialKeyword when (DoesMethodHaveReturnValueOrOutKeyword(currentNode)):
+					case SyntaxKind.PartialKeyword when (PartialMethodRequiresAccessibilityModifier(currentNode)):
 						return;
 					case SyntaxKind.InternalKeyword:
 						if (IsTopLevel(currentNode))
@@ -67,7 +67,7 @@ namespace WTG.Analyzers
 			}
 		}
 
-		static bool DoesMethodHaveReturnValueOrOutKeyword(SyntaxNode node)
+		static bool PartialMethodRequiresAccessibilityModifier(SyntaxNode node)
 		{
 			if (!node.IsKind(SyntaxKind.MethodDeclaration))
 			{
@@ -75,13 +75,16 @@ namespace WTG.Analyzers
 			}
 
 			var methodNode = (MethodDeclarationSyntax)node;
-			if (methodNode.ReturnType is PredefinedTypeSyntax predefinedTypeSyntax
-				&& !predefinedTypeSyntax.Keyword.IsKind(SyntaxKind.VoidKeyword))
+			if (methodNode.ReturnType.IsKind(SyntaxKind.PredefinedType))
 			{
-				return true;
+				var returnType = (PredefinedTypeSyntax)methodNode.ReturnType;
+				if (!returnType.Keyword.IsKind(SyntaxKind.VoidKeyword))
+				{
+					return true;
+				}
 			}
 
-			if (methodNode.ParameterList?.Parameters.Any(c => c.Modifiers.Any(SyntaxKind.OutKeyword)) == true)
+			if (methodNode.ParameterList?.Parameters.Any(static p => p.Modifiers.Any(SyntaxKind.OutKeyword)) == true)
 			{
 				return true;
 			}

--- a/src/WTG.Analyzers/Analyzers/Visibility/VisibilityAnalyzer.cs
+++ b/src/WTG.Analyzers/Analyzers/Visibility/VisibilityAnalyzer.cs
@@ -1,6 +1,8 @@
 using System.Collections.Immutable;
+using System.Linq;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Diagnostics;
 using WTG.Analyzers.Utils;
 
@@ -32,7 +34,8 @@ namespace WTG.Analyzers
 				return;
 			}
 
-			var list = ModifierExtractionVisitor.Instance.Visit(context.Node);
+			var currentNode = context.Node;
+			var list = ModifierExtractionVisitor.Instance.Visit(currentNode);
 			var privateToken = default(SyntaxToken);
 
 			foreach (var modifier in list)
@@ -47,11 +50,10 @@ namespace WTG.Analyzers
 
 					case SyntaxKind.ProtectedKeyword:
 					case SyntaxKind.PublicKeyword:
-					case SyntaxKind.PartialKeyword when (context.Node.IsKind(SyntaxKind.MethodDeclaration)):
+					case SyntaxKind.PartialKeyword when (DoesMethodHaveReturnValueOrOutKeyword(currentNode)):
 						return;
-
 					case SyntaxKind.InternalKeyword:
-						if (IsTopLevel(context.Node))
+						if (IsTopLevel(currentNode))
 						{
 							context.ReportDiagnostic(Rules.CreateDoNotUseTheInternalKeywordForTopLevelTypesDiagnostic(modifier.GetLocation()));
 						}
@@ -63,6 +65,28 @@ namespace WTG.Analyzers
 			{
 				context.ReportDiagnostic(Rules.CreateDoNotUseThePrivateKeywordDiagnostic(privateToken.GetLocation()));
 			}
+		}
+
+		static bool DoesMethodHaveReturnValueOrOutKeyword(SyntaxNode node)
+		{
+			if (!node.IsKind(SyntaxKind.MethodDeclaration))
+			{
+				return false;
+			}
+
+			var methodNode = (MethodDeclarationSyntax)node;
+			if (methodNode.ReturnType is PredefinedTypeSyntax predefinedTypeSyntax
+				&& !predefinedTypeSyntax.Keyword.IsKind(SyntaxKind.VoidKeyword))
+			{
+				return true;
+			}
+
+			if (methodNode.ParameterList?.Parameters.Any(c => c.Modifiers.Any(SyntaxKind.OutKeyword)) == true)
+			{
+				return true;
+			}
+
+			return false;
 		}
 
 		static bool IsTopLevel(SyntaxNode node)

